### DISCUSSION TITLE
StreamingResponse start download while waiting for data

### DISCUSSION THREAD

**Discussion by valeriodeblasio-1** [encode/starlette#2622] [Q&A]
I want to send a file to a user, however the file is generated asynchronously via a thread which can take a long time (even 30 seconds).

I would like the browser to show the file being downloaded right away (with 0 bytes and unknown size), tried via cUrl and it works but with a browser (Chrome or Firefox) this doesn't happen, the download only starts after the first bytes have been sent.

The only way I've been able to find is to send empty characters (\u200B) while waiting for the file to be generated (as suggested in this similar problem [Chrome doesn't show file as being downloaded until 8 bytes are sent (Firefox does)](https://stackoverflow.com/questions/52913925/chrome-doesnt-show-file-as-being-downloaded-until-8-bytes-are-sent-firefox-doe)), but this way the resulting file contains these characters at the beginning, which then gives problems when loading on other systems.

Another method is to use **"Content-Type": "application/force-download"**, but it works only on Chrome (and as is not a standard MIME I don't know if it will broke something)

this is an example code

``` python
import asyncio
from urllib.parse import quote

from starlette.applications import Starlette
from starlette.responses import StreamingResponse, JSONResponse
from starlette.routing import Route


async def async_read_file():
    # this code simulate the activate waiting for the file to be created
    count = 0
    while count < 5:
        await asyncio.sleep(2)
        count += 1

    file = open('file.zip', 'rb')
    while True:
        data = file.read(64 * 1024)
        if not data:
            break
        yield data


async def download(request):
    async def iterate_file_():
        async for chunk in async_read_file():
            yield chunk

    response = StreamingResponse(iterate_file_(), headers={
        "Content-Disposition": "attachment; filename*=utf-8''{}".format(quote('file.zip'))})

    return response



routes = [
    Route("/download", endpoint=download)
]

app = Starlette(debug=True, routes=routes)
```

**Comment by nggit** [comment]
This won't work.

I confirmed that most browsers handle the download dialog after receiving some **real content** (excluding markup like chunked encoding). Even fairly large [chunk extensions](https://datatracker.ietf.org/doc/html/rfc9112#name-chunk-extensions) are ignored, so they expect real content, not just bytes.
